x-pack/plugin/esql/qa/testFixtures/src/main/resources/union_types.csv-spec

```diff
@@ -44,6 +44,44 @@ FROM sample_data_ts_long
 2023-10-23T12:15:03.360Z  |  172.21.2.162  |  3450233              |  Connected to 10.1.0.3
 ;
 
+singleIndexTsNanosAsMillis
+required_capability: to_date_nanos
+
+FROM sample_data_ts_nanos
+| EVAL @timestamp = TO_DATETIME(@timestamp)
+| KEEP @timestamp, client_ip, event_duration, message
+| SORT @timestamp DESC
+;
+
+@timestamp:date           |  client_ip:ip  |  event_duration:long  |  message:keyword
+2023-10-23T13:55:01.543Z  |  172.21.3.15   |  1756467              |  Connected to 10.1.0.1
+2023-10-23T13:53:55.832Z  |  172.21.3.15   |  5033755              |  Connection error
+2023-10-23T13:52:55.015Z  |  172.21.3.15   |  8268153              |  Connection error
+2023-10-23T13:51:54.732Z  |  172.21.3.15   |  725448               |  Connection error
+2023-10-23T13:33:34.937Z  |  172.21.0.5    |  1232382              |  Disconnected
+2023-10-23T12:27:28.948Z  |  172.21.2.113  |  2764889              |  Connected to 10.1.0.2
+2023-10-23T12:15:03.360Z  |  172.21.2.162  |  3450233              |  Connected to 10.1.0.3
+;
+
+singleIndexTsMillisAsNanos
+required_capability: to_date_nanos
+
+FROM sample_data
+| EVAL @timestamp = TO_DATE_NANOS(@timestamp)
+| KEEP @timestamp, client_ip, event_duration, message
+| SORT @timestamp DESC
+;
+
+@timestamp:date_nanos     |  client_ip:ip  |  event_duration:long  |  message:keyword
+2023-10-23T13:55:01.543Z  |  172.21.3.15   |  1756467              |  Connected to 10.1.0.1
```

Comment:
I was expecting here a bunch if zeros after the decimal dot.
I see them [here](https://github.com/elastic/elasticsearch/pull/116265/files#diff-1e8ef4331fa3579f48f3bcdfc034c14a54c0b39b154aa4e4f6c765f21906ab9dR613) and it seems right.

Reply:
This happens because the date nanos [expected values parsing](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestUtils.java#L469) uses the [ISO_DATE_WITH_NANOS](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/util/DateUtils.java#L60) saved format, which specifies `.appendFraction(NANO_OF_SECOND, 3, 9, true)`.  The `3, 9` there specifies a minimum "width" (i.e. number of digits) of 3 and a maximum of 9, so either format would be acceptable.

That said, it's easy enough to add in the zeros, I'll push that up in a moment.  While I was looking at it, I noticed that when I wrote the value parsing logic, I missed an opportunity to use the library function for it, so I've replaced my math with that.  This gets us a little extra error checking for expected values.